gammapy/cube/fit.py

```diff
@@ -208,22 +209,37 @@ def apply_psf(self, npred):
         """Convolve npred cube with PSF"""
         return npred.convolve(self.psf)
 
-    def apply_edisp(self, data):
-        """Convolve map data with energy dispersion."""
-        data = np.rollaxis(data, 0, 3)
+    def apply_edisp(self, npred):
+        """Convolve map data with energy dispersion.
+
+        npred: npred map in e_true
+        Returns
+        -------
+        npred_reco: npred map in e_reco
+        """
+
+        loc = npred.geom.get_axes_dimension_by_name("energy")
+        data = np.moveaxis(npred.data, loc, -1)
         data = np.dot(data, self.edisp.pdf_matrix)
-        return np.rollaxis(data, 2, 0)
+        data1 = np.moveaxis(data, -1, loc) #now dim is in ereco
+        npred1 = Map.from_geom(self.background.geom, unit="")
+        npred1.data = data1
+        return npred1
 
     def compute_npred(self):
-        """Evaluate model predicted counts.
+        """
+        Evaluate model predicted counts.
+        Returns
+        -------
+        npred.data : numpy array
+            array of the predicted counts in each bin (in reco energy)
         """
         flux = self.compute_flux()
         npred = self.apply_exposure(flux)
         if self.psf is not None:
             npred = self.apply_psf(npred)
-        # TODO: discuss and decide whether we need to make map objects in `apply_aeff` and `apply_psf`.
         if self.edisp is not None:
-            npred.data = self.apply_edisp(npred.data)
+            npred = self.apply_edisp(npred)
```

Comment:
I think that's a good change, to always return Map objects now.
If this is a significant inefficiency we can change again later, but probably it's nicer / cleaner like this.